frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/GpuAllocationManager.java

```diff
@@ -51,70 +59,89 @@ public synchronized void addCandidateRequestNode(Node candidateRequestNode) {
-  // To improve it, considers the GPU topology structure, find a node which can minimize
-  // the communication cost between GPUs;
-  public synchronized Node allocateCandidateRequestNode(ResourceDescriptor request, String nodeLabel, String nodeGpuType) {
+  // To improve it, considers the Gpu topology structure, find a node which can minimize
+  // the communication cost between Gpus;
+  public synchronized GpuAllocation selectCandidateRequestNode(ResourceDescriptor request, String nodeLabel, String nodeGpuType) {
     LOGGER.logInfo(
-        "allocateCandidateRequestNode: Request Resource: [%s], NodeLabel: [%s], NodeGpuType: [%s]",
+        "selectCandidateRequestNode: Request Resource: [%s], NodeLabel: [%s], NodeGpuType: [%s]",
         request, nodeLabel, nodeGpuType);
 
     // ClusterConfiguration is ready when this method is called, i.e. it is not null here.
     ClusterConfiguration clusterConfiguration = am.getClusterConfiguration();
     Map<String, NodeConfiguration> nodes = clusterConfiguration.getNodes();
 
-    Iterator<Map.Entry<String, Node>> iter = candidateRequestNodes.entrySet().iterator();
-    Node candidateNode = null;
-    long candidateSelectGPU = 0;
+    GpuAllocation result = null;
 
+    long candidateSelectGpu = 0;
+    Iterator<Map.Entry<String, Node>> iter = candidateRequestNodes.entrySet().iterator();
     while (iter.hasNext()) {
       Map.Entry<String, Node> entry = iter.next();
-      LOGGER.logInfo(
-          "allocateCandidateRequestNode: Try node: " + entry.getValue().toString());
+      LOGGER.logDebug(
+          "selectCandidateRequestNode: Try node: " + entry.getValue().toString());
 
       if (nodeLabel != null) {
         Set<String> nodeLabels = entry.getValue().getNodeLabels();
-        if (!nodeLabels.contains(nodeLabel)) {
-          LOGGER.logInfo(
-              "allocateCandidateRequestNode: Skip node %s, label does not match:%s",
-              entry.getValue().getHostName(), nodeLabel);
+        if (nodeLabels != null && nodeLabels.size() > 0 && !nodeLabels.contains(nodeLabel)) {
+          LOGGER.logDebug(
+              "selectCandidateRequestNode: Skip node %s (nodeLabels: %s), label does not match: request label: %s",
+              entry.getValue().getHostName(), CommonExtensions.toString(nodeLabels), nodeLabel);
           continue;
         }
       }
+      if (nodeGpuType != null) {
+        if (nodes.size() > 0) {
+          if (!nodes.containsKey(entry.getValue().getHostName())) {
+            LOGGER.logWarning(
+                "selectCandidateRequestNode: Skip node %s, getGpuType not set",
+                entry.getValue().getHostName());
+            continue;
+          }
+          String gpuType = (nodes.get(entry.getValue().getHostName())).getGpuType();
+          if (!nodeGpuType.equals(gpuType)) {
+            LOGGER.logDebug(
+                "selectCandidateRequestNode: Skip node %s (gpuType:%s), gpuType don't match: request gpuType: %s",
```

Comment:
space

Reply:
space what ?

---
In reply to: [158621176](https://github.com/Microsoft/pai/pull/128#discussion_r158621176) [](ancestors = 158621176)

Reply:
add space:
(gpuType:%s)
->
(gpuType: %s)

---
In reply to: [158623074](https://github.com/Microsoft/pai/pull/128#discussion_r158623074) [](ancestors = 158623074,158621176)